src/fs/feature/git.rs

```diff
@@ -97,9 +105,15 @@ impl FromIterator<PathBuf> for GitCache {
 
 /// A **Git repository** is one we’ve discovered somewhere on the filesystem.
 pub struct GitRepo {
-    /// The queryable contents of the repository: either a `git2` repo, or the
-    /// cached results from when we queried it last time.
-    contents: Mutex<GitContents>,
+    /// All the interesting Git stuff goes through this.
+    repo: Mutex<git2::Repository>,
+
+    /// Cached path->status mapping.
+    statuses: RwLock<Option<GitStatuses>>,
+
+    /// Cached list of the relative paths of all submodules in this repository.
+    /// This is used to optionally ignore submodule contents when listing recursively.
+    relative_submodule_paths: RwLock<Option<Result<Vec<PathBuf>, git2::Error>>>,
```

Comment:
nit: maybe type alias here? `type SubmodulePaths = Result<Vec<PathBuf>, git2::Error>`
so we can have 
```rust
relative_submodule_paths: RwLock<Option<SubmodulePaths>>,
```